### PR TITLE
stylo: support background-size and corner gradients

### PR DESCRIPTION
**Pull request by Manishearth**
Forgot to fix this in the midst of all other things fixed in #12945

<s>(not yet tested, need to figure out how to test bg-size with gradients)</s>

r? @bholley

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13157)

### CODE REVIEW

**Comment by Manishearth** [comment]
Verified that this works

screen shot 2016-09-01 at 7 51 41 pm

File: components/style/gecko_conversions.rs [hunk 1]
```diff
@@ -70,6 +70,29 @@ impl From<LengthOrPercentage> for nsStyleCoord_CalcValue {
     }
 }
 
+impl LengthOrPercentageOrAuto {
+    pub fn into_calc_value(&self) -> Option<nsStyleCoord_CalcValue> {
```
Comment (bholley) [inline components/style/gecko_conversions.rs]: Is "into" really the right name here? It seems like that would imply that the value is consumed, which it isn't here.

In general it seems like we want to impl From, but maybe that runs into coherence issues? I'd go with as_ or to_ for now.

**Comment by bholley** [comment]
Looks like Simon reviewed the other background array properties, so I'll kick review over to him. :-)

r? @SimonSapin

**Comment by Manishearth** [comment]
Turns out that I was testing with pure gecko. Updating m-c had pulled in changes that required --enable-stylo, which wasn't set.

background-size still works, but it turns out that diagonal gradients do not.

**Comment by Manishearth** [comment]
Fixed corner gradients since it was pretty straightforward. I can split it out into another PR if you want.

screen shot 2016-09-02 at 8 28 13 pm

File: components/style/properties/gecko.mako.rs [hunk 2]
```diff
@@ -1028,6 +1028,75 @@ fn static_assert() {
...
+    pub fn clone_background_size(&self) -> longhands::background_size::computed_value::T {
+        use gecko_bindings::structs::nsStyleCoord_CalcValue as CalcValue;
+        use gecko_bindings::structs::nsStyleImageLayers_Size_DimensionType as DimensionType;
+        use properties::longhands::background_size::single_value::computed_value::{ExplicitSize, T};
+        use values::computed::LengthOrPercentageOrAuto;
+
+        fn to_servo(value: CalcValue, ty: u8) -> LengthOrPercentageOrAuto {
+            if ty == DimensionType::eAuto as u8 {
+                LengthOrPercentageOrAuto::Auto
+            } else {
```
Comment (emilio) [inline components/style/properties/gecko.mako.rs]: `assert!` here?

File: components/style/properties/gecko.mako.rs [hunk 3]
```diff
@@ -1028,6 +1028,75 @@ fn static_assert() {
...
+
+        longhands::background_size::computed_value::T(
+            self.gecko.mImage.mLayers.iter().map(|ref layer| {
+                if DimensionType::eCover as u8 == layer.mSize.mWidthType {
+                    return T::Cover
+                }
+                if DimensionType::eContain as u8 == layer.mSize.mWidthType {
+                    return T::Contain
+                }
+
```
Comment (emilio) [inline components/style/properties/gecko.mako.rs]: Do we want to `assert!` here what `DimensionType` we expect too? (Not totally sure if we can).

**Comment by emilio** [comment]
r=me with the proper assertions added.

File: components/style/properties/gecko.mako.rs (hunk 3 above)
Comment (Manishearth) [inline components/style/properties/gecko.mako.rs]: If we assert in `to_servo` we don't need to again here

Comment (emilio) [inline components/style/properties/gecko.mako.rs]: fair enough.

**Comment by Manishearth** [comment]
asserts added

**Comment by emilio** [comment]
We tempt to prefer `debug_assert!` in general, but well, I guess normal `assert!` is fine too :-)

**Comment by Manishearth** [comment]
heh, I put an assert because I thought you said so -- I prefer debug asserts too (esp in situations where safety isn't important). FIxed

@bors-servo r=emilio

**Comment by Manishearth** [comment]
@bors-servo retry

 - #13059 
- #13165

**Comment by Manishearth** [comment]
@bors-servo retry

 - #13059

**Pull request merged**